src/tests/test_raft_repl_dev.cpp

```diff
@@ -331,45 +331,132 @@ TEST_F(RaftReplDevTest, All_Append_Fetch_Remote_Data) {
 
     g_helper->sync_for_verify_start();
 
-    // TODO: seems with filip and fetch remote, the data size is not correct;
     LOGINFO("Validate all data written so far by reading them");
     this->validate_all_data();
+
     g_helper->sync_for_cleanup_start();
 }
 
-TEST_F(RaftReplDevTest, All_ReplService) {
+// do some io before restart;
+TEST_F(RaftReplDevTest, All_restart_one_follower_inc_resync) {
     LOGINFO("Homestore replica={} setup completed", g_helper->replica_num());
     g_helper->sync_for_test_start();
-    auto repl_dev = dynamic_pointer_cast< RaftReplDev >(pick_one_db().repl_dev());
-    auto group_id = repl_dev->group_id();
-    auto my_id_str = repl_dev->my_replica_id_str();
 
-    auto leader = repl_dev->get_leader_id();
-    ASSERT_TRUE(leader != replica_id_t())
-        << "Error getting leader id for group_id=" << boost::uuids::to_string(group_id).c_str();
-    auto leader_str = boost::uuids::to_string(leader);
-    LOGINFO("Got raft leader {} for group {}", leader_str, group_id);
+    // step-0: do some IO before restart one member;
+    uint64_t exp_entries = 20;
+    if (g_helper->replica_num() == 0) {
+        g_helper->runner().set_num_tasks(20);
+        auto block_size = SISL_OPTIONS["block_size"].as< uint32_t >();
+        LOGINFO("Run on worker threads to schedule append on repldev for {} Bytes.", block_size);
+        g_helper->runner().set_task([this, block_size]() {
+            this->generate_writes(block_size /* data_size */, block_size /* max_size_per_iov */);
+        });
+        g_helper->runner().execute().get();
+    }
 
+    // step-1: wait for all writes to be completed
+    this->wait_for_all_writes(exp_entries);
+
+    // step-2: restart one non-leader replica
+    if (g_helper->replica_num() == 1) {
+        LOGINFO("Restart homestore: replica_num = 1");
+        g_helper->restart();
+        g_helper->sync_for_test_start();
+    }
+
+    exp_entries += SISL_OPTIONS["num_io"].as< uint64_t >();
+    // step-3: on leader, wait for a while for replica-1 to finish shutdown so that it can be removed from raft-groups
+    // and following I/O issued by leader won't be pushed to relica-1;
     if (g_helper->replica_num() == 0) {
-        ASSERT_TRUE(leader_str == my_id_str)
-            << "Leader id " << leader_str.c_str() << " should  equals to my ID " << my_id_str.c_str();
-    } else {
-        ASSERT_TRUE(leader_str != my_id_str) << "I am a follower, Leader id " << leader_str.c_str()
-                                             << " should not equals to my ID " << my_id_str.c_str();
+        LOGINFO("Wait for grpc connection to replica-1 to expire and removed from raft-groups.");
+        std::this_thread::sleep_for(std::chrono::seconds{5});
+
+        g_helper->runner().set_num_tasks(SISL_OPTIONS["num_io"].as< uint64_t >());
+
+        // before replica-1 started, issue I/O so that replica-1 is lagging behind;
+        auto block_size = SISL_OPTIONS["block_size"].as< uint32_t >();
```

Comment:
Wont replica1 restart by the time we wait for 5seconds causing replica1 never lagging behind. Will shutdown without remove files and start homestore of replica1 simulate that ?

Reply:
I think Sanal is right, that we might restart replica-1, which might have resynced by the time the sleep is completed and thus would be a regular append. If so, we can improve the restart method to simulate a sleep between shutdown and restart in (restart_homestore method), which we can control to make sure replica-1 is down while this happens.

Reply:
The shutdown actually takes some time to cmoplete, that is why we sleep 5 seconds here. 
It can be proved everytime replica-1 will be lagging with more than 64 append entries. 
Without this 5 seconds, it needs to do more than 10K I/Os which is fine but unnecessary and verification will also take time to increase the test case run time.

Reply:
There is a 10s sleep between shutdown and start in `start_homestore` so there is no possibility the replica1 started before starting IO(i.e  not possible never lagging behind).

But seems we want to simulate resync while IO are still going on,  this is harder as we are expecting the 300 IOs span longer than (10-5) = 5s.   Better we can continue the IO by triggering one more batch all the way till we see replica1 had joined back, and optionally do one more round of IO. then goes to `this->wait_for_all_writes(exp_entries);`

Reply:
I tested 3000 IOs, and we can actually do that by default, but I thought we don't want to increase the time for homestore build for everyone. I think we better do it in long duration test: https://github.com/eBay/HomeStore/issues/319